### PR TITLE
Trying to pick free connection in order so that we prioritize connect…

### PR DESCRIPTION
**Pull request by godmodegalactus**
…ion that already have been established

### CODE REVIEW

**Review by grooviegermanikus — APPROVED** [review]
hard to say but lgtm

File: core/src/quic_connection.rs [hunk 1]
```diff
@@ -227,21 +226,32 @@ impl QuicConnectionPool {
...
         drop(_others);
+        (permit, index)
+    }
 
+    pub async fn get_pooled_connection(&self) -> anyhow::Result<PooledConnection> {
+        let (permit_result, index) = self.get_permit_and_index().await;
+        let permit = permit_result?;
         // establish a connection if the connection has not yet been used
         let connection = self.connections[index].clone();
         if !connection.has_connected_atleast_once() {
```
[the remark this answers is not in the document]
Comment (godmodegalactus) [reply core/src/quic_connection.rs:252]: We recreate connection but yeah i agree should do that too.

File: core/src/quic_connection.rs [hunk 2]
```diff
@@ -227,21 +226,32 @@ impl QuicConnectionPool {
         }
     }
 
-    pub async fn get_pooled_connection(&self) -> anyhow::Result<PooledConnection> {
+    async fn get_permit_and_index(&self) -> (Result<OwnedSemaphorePermit, AcquireError>, usize) {
```
[the remark this answers is not in the document]
Comment (godmodegalactus) [reply core/src/quic_connection.rs:237]: Only used here.

File: core/src/quic_connection.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (godmodegalactus) [reply core/src/quic_connection.rs:263]: fixed.

File: core/src/quic_connection.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (godmodegalactus) [reply core/src/quic_connection.rs]: fixed.

File: core/src/quic_connection.rs [hunk 3]
```diff
@@ -227,21 +226,32 @@ impl QuicConnectionPool {
...
+            }
+        }
+        // if all of the connections are full then fall back on semaphore which gets free first.
         let (permit, index, _others) = futures::future::select_all(
             self.transactions_in_sending_semaphore
                 .iter()
                 .map(|x| x.clone().acquire_owned().boxed()),
         )
         .await;
         drop(_others);
```
[the remark this answers is not in the document]
Comment (godmodegalactus) [reply core/src/quic_connection.rs]: fixed.

**Review by musitdev — APPROVED** [review]
Ok

**Pull request merged**